### PR TITLE
Add rotation origin variable

### LINKED ISSUE #2734: [Animation System] Rotation around specified origin does not work
I've tried to use the Model Animation System to animate the opening/closing of a door for one of my blocks. There doesn't seem to be a way to set the origin within the animation clips (like you can with the axis), so I added a rotation tag with an angle of 0 to the part I'm rotating instead, which does seem to be loaded [here](https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java#L387).

```
{
    "from": [ 2, 0, 1 ],
    "to": [ 14, 11, 2 ],
    "rotation": { "origin": [0, -8, 8], "axis": "y", "angle": 0 },
    "faces": { ... }
}
```

However, the result of that is this:

origin

It seems to be translating by the origin, rotating, but then not translating back like it's normally done when rotating around a non-centered point. Note how the door is rendering correctly without an offset in the unanimated inventory.

I'm pretty much using the AnimationModelDebug code with some values changed at the moment, since I was just getting started with the new system. Oh, and this is on 1.9.

[Animation State Machine](http://pastebin.com/SPEWrE1r)
[Armature](http://pastebin.com/9ynUMuXx)
+ default AnimationTESR & AnimationStateMachine.load()

### PR DESCRIPTION
**Pull request by mincrmatt12**
Adds an animation variable for setting the rotation point.

Closes #2734.

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java [hunk 1]
```diff
@@ -390,11 +391,25 @@ public TRSRTransformation apply(float time)
...
                     }
                 }
                 Quat4f rot = new Quat4f();
                 rot.set(rotation);
-                return TRSRTransformation.blockCenterToCorner(new TRSRTransformation(translation, rot, scale, null));
+                TRSRTransformation base = new TRSRTransformation(translation, rot, scale, null);
+                Vector3f negOrigin = (Vector3f) origin.clone();
+                negOrigin.negate();
+                base = new TRSRTransformation(origin, null, null, null).compose(base);
+                base = base.compose(new TRSRTransformation(negOrigin, null, null, null));
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java]: you can just combine this with the line above it

File: src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java [hunk 2]
```diff
@@ -390,11 +391,25 @@ public TRSRTransformation apply(float time)
...
+                        case ZORIGIN:
+                            origin.z = value;
+                            break;
                     }
                 }
                 Quat4f rot = new Quat4f();
                 rot.set(rotation);
-                return TRSRTransformation.blockCenterToCorner(new TRSRTransformation(translation, rot, scale, null));
+                TRSRTransformation base = new TRSRTransformation(translation, rot, scale, null);
+                Vector3f negOrigin = (Vector3f) origin.clone();
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java]: Vector3f has a copy method or copy constructor I'm pretty sure which is better style than cloning

Comment (mincrmatt12) [reply src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java]: ok

File: src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java (hunk 1 above)
Comment (mincrmatt12) [reply src/main/java/net/minecraftforge/client/model/animation/ModelBlockAnimation.java]: done

**Comment by mezz** [comment]
Looks good.

Please make a test model that uses this new feature, then I can test it and it can serve as a basic example.

@RainWarrior can you take a look at this? It's very straightforward I think.

**Comment by RainWarrior** [comment]
Yup, looks good to me. Assuming there's not a 3rd place somewhere in code that needs to know about the added parameter (I don't think there is), this should work fine.

**Comment by mincrmatt12** [comment]
It needed a -0.5 on the origin variables because the rotation axes are normally in the center, not the corner.
Also I added another block to the debug mod which uses it

**Comment by mezz** [comment]
The spinning yellow part of the model is not exactly centered (left to right in the screenshot), I assume that is an error somewhere:

I teleported on top of the center of the block, so my crosshair looking down is the exact center.

**Comment by mezz** [comment]
@mincrmatt12 any response?

**Comment by mincrmatt12** [comment]
Sorry i've been super sick the past few weeks, just fixed it @mezz. It was a simple I can't make json models type error

**Comment by mezz** [comment]
Ok, tested and looks good to me.

**Comment by mezz** [comment]
@mincrmatt12 please create this PR for 1.12

**Comment by mincrmatt12** [comment]
Ok, writing one right now

**Comment by wiiv** [comment]
Hello, I've been playing around with this and I'd like to know if the "origin" variables are already implemented for 1.12.

**Comment by mezz** [comment]
This has not been ported to 1.12. 
If you want to do it and open a PR, I think it would be accepted pretty swiftly.

**Comment by williewillus** [comment]
This is just Forge code changes right? It should be almost cleanly cherry-pickable to 1.12, I think, since no patches are involved and this doesn't touch any deprecated stuff

**Comment by mezz** [comment]
Unfortunately there are some conflicts in ModelAnimationDebug.

**Comment by wiiv** [comment]
I'm having trouble matching the origin of the block elements in the model file with the origin I set in the armature file. For example, if I set the origin of the element to be:

"rotation": { "origin": [0, 0, 10], "axis": "y", "angle": 0 }

Which matches the origin on the armature:

{
"variable": "origin_z",
"type": "uniform",
"interpolation": "nearest",
"samples": [ 0.625 ]
}

it just resets the placement of the animated element to [0, 0, 0] and the animation works correctly but not at [0, 0, 10] where it should be happening.

From what I can tell, rotations seem to always be happening around an uncentered point unless they happen at [0, 0, 0].

**Comment by mincrmatt12** [comment]
Does it work if you remove the origin tag from the element in the model file? I never tested what happens when you combine the two.

**Comment by wiiv** [comment]
If I do that the the element starts and ends where I want it but the animation happens around an uncentered point.

**Comment by mincrmatt12** [comment]
Does it work if you remove the origin tag from the element in the model file? I never tested what happens when you combine the two.

**Comment by wiiv** [comment]
I just answered to that.

**Comment by mincrmatt12** [comment]
Can you put in a gist/pastebin your entire armature file? Also you should be defining all rotation related things in the armature file, so there should be no rotation tag on anything in the model file. (also sorry about posting twice, github seems to have hiccuped)

**Comment by wiiv** [comment]
I've put together two GIFs, in the first one I have set the origin in the .json model and this happens (intended movement, wrong placement):
http://imgur.com/A8QjGi9
If I only set the origin on the armature, this happens (wrong movement, intended placement):
http://imgur.com/a/CCuVn

Files:
model file: https://pastebin.com/b2QbuRiZ
armature file: https://pastebin.com/azhPW6Fh

**Comment by wiiv** [comment]
Any updates?

**Comment by mincrmatt12** [comment]
1.12: #4466
@wiiv I think your problem was that you shouldn't specify any rotation in the model at all.
(sorry for the massive delay)

**Pull request merged**